tests/external_user_manager_test.go

```diff
@@ -0,0 +1,347 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/dotmesh-io/dotmesh/pkg/client"
+	"github.com/dotmesh-io/dotmesh/pkg/types"
+	"github.com/dotmesh-io/dotmesh/pkg/user"
+
+	"github.com/dotmesh-io/citools"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type DummyUserManager struct {
+	allowStuff       bool
+	theValidAdminKey string
+	theValidKey      string
+	log              log.FieldLogger
+}
+
+func (m *DummyUserManager) NewAdmin(user *user.User) error {
+	m.log.Infof("NewAdmin: %#v (%s)", *user, string(user.Password))
+	m.theValidAdminKey = user.ApiKey
+	return nil
+}
+
+func (m *DummyUserManager) New(name, email, password string) (*user.User, error) {
+	m.log.Infof("New: %q %q %q", name, email, password)
+	return &user.User{
+		Id:       email,
+		Name:     name,
+		Email:    email,
+		Salt:     []byte{},
+		Password: []byte(password),
+		ApiKey:   "123",
+	}, nil
+}
+func (m *DummyUserManager) Get(q *user.Query) (*user.User, error) {
+	m.log.Infof("Get: %q %q", q.Ref, q.Selector)
+	return &user.User{
+		Id:   q.Ref,
+		Name: q.Ref,
+	}, nil
+}
+
+func (m *DummyUserManager) Update(user *user.User) (*user.User, error) {
+	m.log.Infof("Update: %#v", *user)
+	return user, nil
+}
+
```

Comment:
I recently discovered whilst writing something in the dotscience-agent that go-mocks is pretty nice to use for unit testing interfaces. Worth checking it out. There are examples in the repo. https://github.com/golang/mock

Reply:
yeah, it's currently my "Go To" thing too, it also tests the ordering of calls and ensures that no unknown calls are made